### PR TITLE
Add `ArrayTrait`

### PR DESCRIPTION
**Pull request by RiscadoA**
# Description

Adds the `ArrayTrait` and defines reflection for `std::vector<T>` types.

## Checklist

- [x] Self-review changes.
- [X] Evaluate impact on the documentation.
- [X] Ensure test coverage.
- [X] Write new samples.

### CODE REVIEW

File: core/include/cubos/core/reflection/traits/array.hpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+        /// @brief Inserts a default-constructed element into the array.
+        /// @note Aborts if @ref hasInsertDefault() returns false.
+        /// @param instance Pointer to the array instance.
+        /// @param index Element index.
+        /// @return Whether the operation is supported.
+        bool insertDefault(void* instance, std::size_t index) const;
+
+        /// @brief Inserts a copy-constructed element into the array.
+        /// @note Aborts if @ref hasInsertCopy() returns false.
```
Comment (RiscadoA) [inline core/include/cubos/core/reflection/traits/array.hpp:99]:
```suggestion
```
Outdated, applies to other inserts and erase.

File: core/include/cubos/core/reflection/traits/array.hpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        /// @param instance Pointer to the array instance.
+        /// @param index Element index.
+        /// @param value Value.
+        /// @return Whether the operation is supported.
+        bool insertCopy(void* instance, std::size_t index, const void* value) const;
+
+        /// @brief Inserts a move-constructed element into the array.
+        /// @note Aborts if @ref hasInsertMove() returns false.
```
Comment (RiscadoA) [inline core/include/cubos/core/reflection/traits/array.hpp:107]:
```suggestion
```

File: core/include/cubos/core/reflection/traits/array.hpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        /// @param instance Pointer to the array instance.
+        /// @param index Element index.
+        /// @param value Value.
+        /// @return Whether the operation is supported.
+        bool insertMove(void* instance, std::size_t index, void* value) const;
+
+        /// @brief Removes an element of the array.
+        /// @note Aborts if @ref hasErase() returns false.
```
Comment (RiscadoA) [inline core/include/cubos/core/reflection/traits/array.hpp:115]:
```suggestion
```

File: core/include/cubos/core/reflection/traits/array.hpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        /// @param instance Pointer to the array instance.
+        /// @param index Element index.
+        /// @return Pointer to the element.
+        void* get(void* instance, std::size_t index) const;
+
+        /// @copydoc get(void*, std::size_t) const
+        const void* get(const void* instance, std::size_t index) const;
+
+        /// @brief Inserts a default-constructed element into the array.
+        /// @note Aborts if @ref hasInsertDefault() returns false.
```
Comment (RiscadoA) [inline core/include/cubos/core/reflection/traits/array.hpp:92]:
```suggestion
```

**Review by luishfonseca — CHANGES_REQUESTED** [review]

File: core/samples/reflection/traits/array/main.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+{
+    const auto& arrayTrait = type.get<ArrayTrait>();
+    /// [Printing any array]
+
+    /// [Getting array length and type]
+    CUBOS_INFO("Array with {} elements of type {}", arrayTrait.length(instance), arrayTrait.elementType().name());
+    /// [Getting array length and type]
+
+    /// [Getting array elements]
+    if (!arrayTrait.elementType().is<int>())
```
Comment (RiscadoA) [inline core/samples/reflection/traits/array/main.cpp:20]:
```suggestion
    if (!arrayTrait.elementType().is<int32_t>())
```

File: core/samples/reflection/traits/array/main.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        CUBOS_INFO("This function does not support printing arrays of types other than int");
+        return;
+    }
+
+    for (std::size_t i = 0; i < arrayTrait.length(instance); ++i)
+    {
+        CUBOS_INFO("Element {}: {}", i, *static_cast<const int*>(arrayTrait.get(instance, i)));
```
Comment (RiscadoA) [inline core/samples/reflection/traits/array/main.cpp:28]:
```suggestion
        CUBOS_INFO("Element {}: {}", i, *static_cast<const int32_t*>(arrayTrait.get(instance, i)));
```

File: core/samples/reflection/traits/array/page.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+To make calling our function easier, we can add a convenience typed wrapper:
+
+@snippet reflection/traits/array/main.cpp Typed wrapper
+
+Using this function is now as simple as:
+
+@snippet reflection/traits/array/main.cpp Usage
+
+Its important to note that both the includes above are necessary, as we're
+reflecting the type `std::vector<int>`, which also means reflecting `int`.
```
Comment (RiscadoA) [inline core/samples/reflection/traits/array/page.md:32]:
```suggestion
reflecting the type `std::vector<int32_t>`, which also means reflecting `int32_t`.
```

File: core/samples/reflection/traits/array/page.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+Through the trait, we can access the size of the array and its element type:
+
+@snippet reflection/traits/array/main.cpp Getting array length and type
+
+We can also get pointers to the elements of the array and iterate over them:
+
+@snippet reflection/traits/array/main.cpp Getting array elements
+
+In this example, we're only supporting arrays of `int`s, but we could for
```
Comment (RiscadoA) [inline core/samples/reflection/traits/array/page.md:20]:
```suggestion
In this example, we're only supporting arrays of `int32_t`s, but we could for
```

File: core/samples/reflection/traits/array/main.cpp [hunk 9]
```diff
@@ -17,15 +17,15 @@ void printArray(const Type& type, const void* instance)
     /// [Getting array length and type]
 
     /// [Getting array elements]
-    if (!arrayTrait.elementType().is<int>())
+    if (!arrayTrait.elementType().is<int32_t>())
     {
         CUBOS_INFO("This function does not support printing arrays of types other than int");
```
Comment (luishfonseca) [inline core/samples/reflection/traits/array/main.cpp:22]:
```suggestion
        CUBOS_INFO("This function does not support printing arrays of types other than int32_t");
```

**Review by luishfonseca — CHANGES_REQUESTED** [review]

**Review by luishfonseca — COMMENTED** [review]
Should we add `.view(...)`? It would allow the following syntax
```cpp
for (auto* value : arrayTrait.view(myArray))
{
...
}
```

**Review by luishfonseca — APPROVED** [review]

**Pull request merged**